config/sampler_config.go

```diff
@@ -45,12 +47,379 @@ type RulesBasedSamplerCondition struct {
+		if err != nil {
+			return err
+		}
+	case "contains":
+		err := setMatchStringBasedOperators(r, "contains")
+		if err != nil {
+			return err
+		}
+	case "does-not-contain":
+		err := setMatchStringBasedOperators(r, "does-not-contain")
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown operator '%s'", r.Operator)
+	}
+	return nil
+}
+
+func tryConvertToInt(v any) (int, bool) {
+	switch value := v.(type) {
+	case int:
+		return value, true
+	case int64:
+		return int(value), true
+	case float64:
+		return int(value), true
+	case bool:
+		return 0, false
+	case string:
+		n, err := strconv.Atoi(value)
+		if err == nil {
+			return n, true
+		}
+		return 0, false
+	default:
+		return 0, false
+	}
+}
+
+func tryConvertToInt64(v any) (int64, bool) {
+	switch value := v.(type) {
+	case int:
+		return int64(value), true
+	case int64:
+		return value, true
+	case float64:
+		return int64(value), true
+	case bool:
+		return 0, false
+	case string:
+		n, err := strconv.ParseInt(value, 10, 64)
+		if err == nil {
+			return n, true
+		}
+		return 0, false
+	default:
+		return 0, false
+	}
+}
+
+func tryConvertToFloat(v any) (float64, bool) {
+	switch value := v.(type) {
+	case float64:
+		return value, true
+	case int:
+		return float64(value), true
+	case int64:
+		return float64(value), true
+	case bool:
+		return 0, false
+	case string:
+		n, err := strconv.ParseFloat(value, 64)
+		return n, err == nil
+	default:
+		return 0, false
+	}
+}
+
+func tryConvertToString(v any) (string, bool) {
+	switch value := v.(type) {
+	case string:
+		return value, true
+	case int:
+		return strconv.Itoa(value), true
+	case int64:
+		return strconv.FormatInt(value, 10), true
+	case float64:
+		return strconv.FormatFloat(value, 'E', -1, 64), false
+	case bool:
+		return strconv.FormatBool(value), true
+	default:
+		return "", false
+	}
+}
+
+func tryConvertToBool(v any) bool {
+	switch value := v.(type) {
+	case string:
+		if value == "true" {
+			return true
+		}
+		return false
+	default:
+		return false
+	}
+}
+
+func setCompareOperators(r *RulesBasedSamplerCondition, condition string) error {
+	switch conditionValue := r.Value.(type) {
+	case string:
+		// check if conditionValue and spanValue are not equal
+		switch condition {
+		case "!=":
+			r.Matches = func(spanValue any, exists bool) bool {
+				if n, ok := tryConvertToString(spanValue); exists && ok {
+					return n != conditionValue
+				}
+				return false
+			}
+			return nil
+		case "=":
+			r.Matches = func(spanValue any, exists bool) bool {
+				if n, ok := tryConvertToString(spanValue); exists && ok {
+					return n == conditionValue
+				}
+				return false
+			}
+			return nil
+		}
+	case int:
+		// check if conditionValue and spanValue are not equal
+		switch condition {
+		case "!=":
+			r.Matches = func(spanValue any, exists bool) bool {
+				if n, ok := tryConvertToInt(spanValue); exists && ok {
+					return n != conditionValue
```

Comment:
All the match functions end up being the same except for the comparator.  I think we may be able to take advantage of some generics/extra functions to reduce the duplicated code.  Take a look at what @kentquirk did with a [similar comparitor problem in OTTL](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/compare.go).

Reply:
Yeah, I did talk about doing something like this with @kentquirk before this project ended up dragging a bit. I think we decided to push this change first and then make that change as a refactor in a second PR.